File: data-structures/dynamic-ring-buffer/tests/dynamic_ring_buffer_test.go
package tests

import (
	"reflect"
	"testing"

	"dsa/data-structures/dynamic-ring-buffer/implementations"
)

func TestDynamicRingBufferSimpleNewBufferStartsEmpty(t *testing.T) {
	buffer := implementations.NewMyDynamicRingBuffer[int](2)

	if got := buffer.Length(); got != 0 {
		t.Fatalf("expected length 0, got %d", got)
	}

	if got := buffer.Capacity(); got != 2 {
		t.Fatalf("expected capacity 2, got %d", got)
	}

	if !buffer.IsEmpty() {
		t.Fatal("expected buffer to be empty")
	}

	if got := buffer.ToSlice(); len(got) != 0 {
		t.Fatalf("expected empty slice, got %v", got)
	}
}

func TestDynamicRingBufferSimpleEnqueueAlwaysSucceeds(t *testing.T) {
	buffer := implementations.NewMyDynamicRingBuffer[int](2)

	if !buffer.Enqueue(10) {
		t.Fatal("expected enqueue of 10 to succeed")
	}
	if !buffer.Enqueue(20) {
		t.Fatal("expected enqueue of 20 to succeed")
	}
	if !buffer.Enqueue(30) {
		t.Fatal("expected enqueue of 30 to succeed")
	}

	if got := buffer.Length(); got != 3 {
		t.Fatalf("expected length 3, got %d", got)
	}

	if got := buffer.Front(); got != 10 {
		t.Fatalf("expected front 10, got %d", got)
	}

	if got := buffer.Rear(); got != 30 {
		t.Fatalf("expected rear 30, got %d", got)
	}

	if got := buffer.ToSlice(); !reflect.DeepEqual(got, []int{10, 20, 30}) {
		t.Fatalf("expected [10 20 30], got %v", got)
	}
}

func TestDynamicRingBufferSimpleDequeueReturnsItemsInFIFOOrder(t *testing.T) {
	buffer := implementations.NewMyDynamicRingBuffer[string](2)

	buffer.Enqueue("a")
	buffer.Enqueue("b")
	buffer.Enqueue("c")

	if got := buffer.Dequeue(); got != "a" {
		t.Fatalf("expected a, got %s", got)
	}

	if got := buffer.Dequeue(); got != "b" {
		t.Fatalf("expected b, got %s", got)
	}

	if got := buffer.Front(); got != "c" {
		t.Fatalf("expected front c, got %s", got)
	}
}

func TestDynamicRingBufferSimpleClearResetsTheBuffer(t *testing.T) {
	buffer := implementations.NewMyDynamicRingBuffer[int](2)

	buffer.Enqueue(1)
	buffer.Enqueue(2)
	buffer.Clear()

	if got := buffer.Length(); got != 0 {
		t.Fatalf("expected length 0, got %d", got)
	}

	if !buffer.IsEmpty() {
		t.Fatal("expected buffer to be empty")
	}
}

func TestDynamicRingBufferMildlyAbsurdGrowthPreservesOrderWhenWrapped(t *testing.T) {
	buffer := implementations.NewMyDynamicRingBuffer[int](3)

	buffer.Enqueue(1)
	buffer.Enqueue(2)
	buffer.Enqueue(3)
	buffer.Dequeue()
	buffer.Enqueue(4)
	buffer.Enqueue(5)

	if got := buffer.ToSlice(); !reflect.DeepEqual(got, []int{2, 3, 4, 5}) {
		t.Fatalf("expected [2 3 4 5], got %v", got)
	}

	if got := buffer.Front(); got != 2 {
		t.Fatalf("expected front 2, got %d", got)
	}

	if got := buffer.Rear(); got != 5 {
		t.Fatalf("expected rear 5, got %d", got)
	}

	if got := buffer.Capacity(); got < 4 {
		t.Fatalf("expected capacity to grow to at least 4, got %d", got)
	}
}

func TestDynamicRingBufferAbsurdDeterministicOperationSequenceMatchesReferenceModel(t *testing.T) {
	buffer := implementations.NewMyDynamicRingBuffer[int](2)
	reference := make([]int, 0)

	for index := 1; index <= 80; index++ {
		buffer.Enqueue(index)
		reference = append(reference, index)

		if index%5 == 0 {
			buffer.Dequeue()
			reference = reference[1:]
		}

		if index%8 == 0 {
			buffer.Enqueue(index * 10)
			reference = append(reference, index*10)
		}
	}

	if got := buffer.ToSlice(); !reflect.DeepEqual(got, reference) {
		t.Fatalf("expected %v, got %v", reference, got)
	}

	if got := buffer.Length(); got != len(reference) {
		t.Fatalf("expected length %d, got %d", len(reference), got)
	}

	if got := buffer.Capacity(); got < len(reference) {
		t.Fatalf("expected capacity to be at least %d, got %d", len(reference), got)
	}
}

File: data-structures/dynamic-ring-buffer/tests/dynamic-ring-buffer.test.ts
import assert from "node:assert/strict";
import { MyDynamicRingBuffer } from "../implementations/dynamic-ring-buffer";

function runTest(name: string, execute: () => void): void {
  try {
    execute();
    console.log(`PASS ${name}`);
  } catch (error) {
    console.error(`FAIL ${name}`);
    throw error;
  }
}

runTest("simple: new buffer starts empty", () => {
  const buffer = new MyDynamicRingBuffer<number>(2);

  assert.equal(buffer.length(), 0);
  assert.equal(buffer.capacity(), 2);
  assert.equal(buffer.isEmpty(), true);
  assert.equal(buffer.isFull(), false);
  assert.equal(buffer.front(), undefined);
  assert.equal(buffer.rear(), undefined);
  assert.deepEqual(buffer.toArray(), []);
});

runTest("simple: enqueue always succeeds and exposes front/rear", () => {
  const buffer = new MyDynamicRingBuffer<number>(2);

  assert.equal(buffer.enqueue(10), true);
  assert.equal(buffer.enqueue(20), true);
  assert.equal(buffer.enqueue(30), true);

  assert.equal(buffer.length(), 3);
  assert.equal(buffer.front(), 10);
  assert.equal(buffer.rear(), 30);
  assert.deepEqual(buffer.toArray(), [10, 20, 30]);
});

runTest("simple: dequeue returns items in FIFO order", () => {
  const buffer = new MyDynamicRingBuffer<string>(2);

  buffer.enqueue("a");
  buffer.enqueue("b");
  buffer.enqueue("c");

  assert.equal(buffer.dequeue(), "a");
  assert.equal(buffer.dequeue(), "b");
  assert.equal(buffer.front(), "c");
  assert.deepEqual(buffer.toArray(), ["c"]);
});

runTest("simple: clear resets the buffer", () => {
  const buffer = new MyDynamicRingBuffer<number>(2);

  buffer.enqueue(1);
  buffer.enqueue(2);
  buffer.clear();

  assert.equal(buffer.length(), 0);
  assert.equal(buffer.isEmpty(), true);
  assert.equal(buffer.dequeue(), undefined);
});

runTest("mildly absurd: growth preserves logical order when the buffer was wrapped", () => {
  const buffer = new MyDynamicRingBuffer<number>(3);

  buffer.enqueue(1);
  buffer.enqueue(2);
  buffer.enqueue(3);
  buffer.dequeue();
  buffer.enqueue(4);
  buffer.enqueue(5);

  assert.deepEqual(buffer.toArray(), [2, 3, 4, 5]);
  assert.equal(buffer.front(), 2);
  assert.equal(buffer.rear(), 5);
  assert.ok(buffer.capacity() >= 4);
});

runTest("mildly absurd: growth keeps happening across many wrap-then-grow cycles", () => {
  const buffer = new MyDynamicRingBuffer<number>(1);
  const reference: number[] = [];

  for (let value = 0; value < 40; value += 1) {
    buffer.enqueue(value);
    reference.push(value);

    if (value % 4 === 0 && reference.length > 0) {
      buffer.dequeue();
      reference.shift();
    }
  }

  assert.deepEqual(buffer.toArray(), reference);
  assert.equal(buffer.length(), reference.length);
  assert.ok(buffer.capacity() >= reference.length);
});

runTest("absurd: deterministic operation sequence matches a reference model", () => {
  const buffer = new MyDynamicRingBuffer<number>(2);
  const reference: number[] = [];

  for (let index = 1; index <= 80; index += 1) {
    buffer.enqueue(index);
    reference.push(index);

    if (index % 5 === 0) {
      buffer.dequeue();
      reference.shift();
    }

    if (index % 8 === 0) {
      buffer.enqueue(index * 10);
      reference.push(index * 10);
    }
  }

  assert.deepEqual(buffer.toArray(), reference);
  assert.equal(buffer.length(), reference.length);
  assert.equal(buffer.front(), reference[0]);
  assert.equal(buffer.rear(), reference[reference.length - 1]);
  assert.ok(buffer.capacity() >= reference.length);
});
